### PR TITLE
Add adjoint for Zip

### LINKED ISSUE #221: Gradient doesn't work properly with `enumerate`, `zip`
`ERROR: Variable %56 in block 4 is not defined.` when using enumerate, zip.

MWE: (on julia-1.2rc1, IRTools master, Zygote master)
```julia
s = rand(10)
d = rand(10)

Zygote.gradient(s) do
       sums = 0
       for (a, b) in zip(s, d)
            sums += a + b
       end
       return sums
end
```

cc @dhairyagandhi

### PR DESCRIPTION
**Pull request by DhairyaLGandhi**
Fixes #221

### CODE REVIEW

**Comment by willtebbutt** [comment]
Is there a reason for this to be implemented here, rather than in `ChainRules`?

**Comment by DhairyaLGandhi** [comment]
We already have a bunch of the infrastructure here already, but I agree we should talk about exporting `rrules`

**Comment by DhairyaLGandhi** [comment]
The resolution error in CI is annoying, did anything change in the Reexport package?

File: src/lib/base.jl [hunk 1]
```diff
@@ -137,6 +137,10 @@ end
     return getfield(p, i), pair_getfield
 end
 
+@adjoint function Base.Iterators.Zip(is)
+  Base.Iterators.Zip(is), Δ -> (unzip(Δ),)
+end
```
Comment (willtebbutt) [inline src/lib/base.jl:142]:
```suggestion
@adjoint function Base.Iterators.Zip(is)
  Zip_pullback(Δ) = (unzip(Δ),)
  return Base.Iterators.Zip(is), Zip_pullback
end
```
so that the `@code_warntype` is readable.

File: test/gradcheck.jl [hunk 2]
```diff
@@ -180,6 +180,11 @@ end
   @test back([nothing]) == ([nothing for _ in 1:3], nothing)
 end
 
+@testset "Zip" begin
```
Comment (willtebbutt) [inline test/gradcheck.jl:183]: I'm slightly nervous that there's only one test here, given that we've got the dependence on `_unzip` which needs the `Zip` to be non-empty and for `length` to be defined over the things being iterated over.

It would be good to add some additional tests on objects without a defined length and empty `Zip`s, since we've got no way to exclude them via dispatch.

**Review by willtebbutt — COMMENTED** [review]
Going to hold off approving because I think that we might need more testing.

**Comment by DhairyaLGandhi** [comment]
Testing is fair, will add a few. I suspect we would end up generalising `unzip` to some extent to support these cases more generally, which seems like a good thing

**Comment by CarloLucibello** [comment]
the issue has been already fixed
